subsys/bluetooth/mesh/sensor_cli.c

```diff
@@ -0,0 +1,942 @@
+	/* A packet with only the sensor ID means that the given sensor doesn't
+	 * exist on the sensor server.
+	 */
+	if (buf->len == 2) {
+		ack_ctx->count = 0;
+		goto yield_ack;
+	}
+
+	struct bt_mesh_sensor_info sensor;
+
+	while (buf->len >= 8) {
+		u32_t tolerances;
+
+		sensor.id = net_buf_simple_pull_le16(buf);
+		tolerances = net_buf_simple_pull_le24(buf);
+		tolerance_decode(tolerances & BIT_MASK(12),
+				 &sensor.descriptor.tolerance.positive);
+		tolerance_decode(tolerances >> 12,
+				 &sensor.descriptor.tolerance.negative);
+		sensor.descriptor.sampling_func = net_buf_simple_pull_u8(buf);
+		sensor.descriptor.period =
+			sensor_powtime_decode(net_buf_simple_pull_u8(buf));
+		sensor.descriptor.update_interval =
+			sensor_powtime_decode(net_buf_simple_pull_u8(buf));
+
+		if (cli->cb && cli->cb->sensor) {
+			cli->cb->sensor(cli, ctx, &sensor);
+		}
+
+		if (is_rsp && count < ack_ctx->count) {
+			ack_ctx->sensors[count++] = sensor;
+		}
+	}
+
+yield_ack:
+	if (is_rsp) {
+		ack_ctx->count = count;
+		model_ack_rx(&cli->ack);
+	}
+}
+
+static void handle_status(struct bt_mesh_model *mod,
+			  struct bt_mesh_msg_ctx *ctx,
+			  struct net_buf_simple *buf)
+{
+	struct bt_mesh_sensor_cli *cli = mod->user_data;
+	struct sensor_data_rsp *rsp = cli->ack.user_data;
+	bool is_rsp;
+	int err;
+
+	is_rsp = model_ack_match(&cli->ack, BT_MESH_SENSOR_OP_STATUS, ctx);
+
+	while (buf->len > 3) {
+		const struct bt_mesh_sensor_type *type;
+		u8_t length;
+		u16_t id;
+
+		sensor_status_id_decode(buf, &length, &id);
+		if (length == 0) {
+			if (is_rsp && rsp->id == id) {
+				rsp->value = NULL;
+				rsp->id = BT_MESH_PROP_ID_PROHIBITED;
+				model_ack_rx(&cli->ack);
+			}
+
+			continue;
+		}
+
+		type = bt_mesh_sensor_type_get(id);
+		if (!type) {
+			net_buf_simple_pull(buf, length);
+			unknown_type(cli, ctx, id, BT_MESH_SENSOR_OP_STATUS);
+			continue;
+		}
+
+		u8_t expected_len = sensor_value_len(type);
+
+		if (length != expected_len) {
+			BT_WARN("Invalid length for 0x%04x: %u (expected %u)",
+				id, length, expected_len);
+			return;
+		}
+
+		struct sensor_value value[CONFIG_BT_MESH_SENSOR_CHANNELS_MAX];
+
+		err = sensor_value_decode(buf, type, value);
+		if (err) {
+			return; /* Invalid format, should ignore message */
+		}
+
+		if (cli->cb && cli->cb->data) {
+			cli->cb->data(cli, ctx, type, value);
+		}
+
+		if (is_rsp && rsp->id == id) {
+			memcpy(rsp->value, value,
+			       sizeof(struct sensor_value) *
+				       type->channel_count);
+			rsp->id = BT_MESH_PROP_ID_PROHIBITED;
+			model_ack_rx(&cli->ack);
+		}
+	}
+}
+
+static int parse_series_entry(const struct bt_mesh_sensor_type *type,
+			      const struct bt_mesh_sensor_format *col_format,
+			      struct net_buf_simple *buf,
+			      struct bt_mesh_sensor_series_entry *entry)
+{
+	int err;
+
+	err = sensor_ch_decode(buf, col_format, &entry->column.start);
+	if (err) {
+		return err;
+	}
+
+	if (buf->len == 0) {
+		BT_WARN("The requested column didn't exist: %s%u.%06u",
```

Comment:
Maybe exstract the string encoding of the value to a public API function that the application could use as well?
Similar to `bt_addr_le_to_str`.